cli/src/alluxio.org/cli/processes/common.go

```diff
@@ -12,113 +12,150 @@
 package processes
 
 import (
-	"bufio"
 	"fmt"
 	"io"
+	"io/ioutil"
 	"os"
-	"path"
+	"path/filepath"
 	"strings"
 	"time"
 
+	"github.com/palantir/stacktrace"
 	"golang.org/x/crypto/ssh"
 
 	"alluxio.org/cli/env"
 	"alluxio.org/log"
 )
 
-func getMasters() ([]string, error) {
-	mastersDir := path.Join(env.Env.EnvVar.GetString(env.ConfAlluxioConfDir.EnvVar), "masters")
-	mastersFile, err := os.Open(mastersDir)
-	if err != nil {
-		log.Logger.Errorf("Error reading worker hostnames at %s", mastersDir)
-		return nil, err
+var cliPath = filepath.Join(env.Env.EnvVar.GetString(env.ConfAlluxioHome.EnvVar), "bin", "alluxio")
+var privateKeySigner ssh.Signer
+var masterList []string
+var workerList []string
+
+// for each master, create a client and run
+// TODO: now start master one by one, need to do them in parallel
+func runCommand(command string, onMasters bool, onWorkers bool) error {
+	// get list of masters and workers
+	var nodes []string
+	if onMasters {
+		if len(masterList) == 0 {
```

Comment:
the current logic is:
- if need master list
  - if master list empty
    - parse masters file
- if need worker list
  - if worker list empty
    - parse workers file

there are several problems with this approach
- problem: how will this logic expand if there is a third file or more?
  - the current logic is copied and pasted for each file
  - solution: use one function with a file name as input and list of hostnames as ouptut
- problem: if there are concurrent (multiple parallel) threads calling this function, it is possible that the resulting list will have duplicates
  - ex. 3 threads simultaneously check that `len(masterList) == 0` and then each append `masterList` -> `masterList` is x3 larger than expected
  - solution: each list needs a lock to ensure only one thread can write to the list. the pattern looks like
```
var mutex := &sync.Mutex{}

func getList() {
  mutex.Lock()
  defer mutex.Unlock()
  if len(list) == 0 {
    // write to list
  }
  return list
}
```
- nit: use sync.Once to run a piece of code exactly once instead of doing a `len(list) == 0` check

however since there needs to be one lock per file, it would not be good to create n locks outside the function. to solve all these problems at once, we can introduce a new handler struct that represents a hostname list file.

```
type HostnamesFile struct {
  Name string // ie. "masters"
  Mutex *sync.Mutex
  Once *sync.Once
  Hostnames []string
}

func NewHostamesFile(name string) *HostnamesFile {
  return &HostnamesFile{
    Name: name,
    Mutex: &sync.Mutex{},
    Once: &sync.Once{},
}

func (f *HostnamesFile) Hostnames() ([]string, error) {
  f.Mutex.Lock()
  defer f.Mutex.Unlock()
  var parseErr error
  f.Once.Do(func() {
    // parse file, setting `parseErr` if encountered an error
    // set f.Hostnames
  })
  if parseErr != nil {
    return nil, stacktrace.Propagate(parseErr, "error parsing hostnames file")
  }
  return f.Hostnames
}
```

initialize a `hostnamesGroup := map[string]*HostnamesFile` with an entry for `masters` and `workers` and then call `hostnamesGroup[name].Hostnames()` to get the hostnames list